content/posts/distributed-business-processes.md

```diff
@@ -0,0 +1,80 @@
+- for each instance of initial state change, the follow-up state change has to occur exactly once
+- the process that executed the follow-up state change has to see the initial state change as done
+
+A necessary element of each distributed business process is communication. The fact that the initial state change has occurred needs to be communicated to the code that executes the follow-up change. Let's see what is required for such communication.
+
+### Reliability
+
+We want the communication to be reliable to make sure that the message that should trigger the follow-up change is not lost.
+
+### De-duplication
+
+We want the messages that trigger the follow-up change to be de-duplicated in such a way that if the follow-up change is already persisted, receiving another copy of the trigger message has no effect.
+
+### Causal order
+
+We want the message that triggers the follow-up change to not be delivered before the initial state change is durably persisted.
+
+## Consistent messaging
+
+We mentioned [consistent messaging](https://exactly-once.github.io/posts/consistent-messaging/) before
+
+>we want an endpoint to produce observable side-effects equivalent to some execution in which each logical message gets processed exactly-once. Equivalent meaning that it’s indistinguishable from the perspective of any other endpoint.
+
+Now we can see that this definition is not enough to guarantee correct distributed business process execution. In addition to the above, we also need causal order and reliable delivery. That translates an additional clause:
+
+>if the side effects include sending out messages, the messages should not be dispatched before other side effects are persisted. Dispatched messages should be stored durably.
+
+## Idempotence is not enough
+
+It is a common misconception that idempotence, as a property of a persistence code, is all that is required to build distributed systems. Well, not when distributed business processes are considered. Let's assume that the state change application is idempotent. That means it can be applied multiple times and the side effects (in this case the state of the data) is exactly the same as if it was applied only once. We will now use that idempotent state change operation in our message handler. The handler is receiving messages and sending out messages to trigger the follow-up state change.
+
+We receive a message and invoke the idempotent state change operation. Now we should send out a messages that informs about the state change. But what state change? We don't know if any state change occurred as part of this execution. All we know is the current state is exactly as if the change occurred exactly once. The only message we can send would carry the current state. Such a message is useless as far as distributed business processes are concerned because it cannot trigger any follow-up changes. The distributed business process cannot continue.
```

Comment:
```suggestion
The handler receives a message and invokes the idempotent state change operation. Now we should send out a message that informs about the state change. But what state change? We don't know if any state change occurred as part of this execution. All we know is the current state is exactly as if the change occurred exactly once. The only message we can send would carry the current state. Such a message is useless as far as distributed business processes are concerned because it cannot trigger any follow-up changes. The distributed business process cannot continue.
```